nixosTests.nixos-test-driver: init

Synthetica9 opened the pull request:
###### Motivation for this change

Added minimal changes to make it properly testable, mostly adding a `timeout` option to functions.

cc @marijanp @tfc 

Minimal version of #154168 


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Synthetica9 committed:
nixosTests.nixos-test-driver: init

Synthetica9 committed:
nixos/test-driver: add and export Driver.must_raise

Synthetica9 committed:
nixos/test-driver: remove magic number for timeout

Synthetica9 committed:
nixos/test-driver: add timeout to wait_for_file

Synthetica9 committed:
nixos/test-driver: add timeout to wait_until_tty_matches

Synthetica9 committed:
nixos/test-driver: add timeout to wait_for_text

Synthetica9 committed:
nixos/test-driver: add timeout to wait_for_window

marijanp commented on nixos/lib/test-driver/run_mutmut.sh:
```diff
@@ -0,0 +1,12 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -I nixpkgs=../../.. -p black python3Packages.mutmut python3Packages.coverage -i bash
```
we could wrap this script using wrapProgram, adding all the runtime-dependencies to the PATH, moreover passing the to be mutated path using an argument, and use wrapPrograms -set to set it to the path.

marijanp commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -32,6 +33,29 @@ def get_tmp_dir() -> Path:
 
+@contextmanager
+def must_raise(
+    regex: str,
```
add a short explanation

marijanp commented on nixos/lib/test-driver/test_driver/machine.py:
```diff
@@ -804,23 +807,26 @@ def screen_matches(last: bool) -> bool:
-                if matches is not None:
-                    return
+                    return False
+                # TODO: call run_callbacks() here
```
I suspect this is for future work

Synthetica9 replied:
Yeah, currently waiting for console text that never shows up breaks the cooperative multitasking model we use for `polling_condition`. Not for this PR tho.

marijanp commented on nixos/tests/nixos-test-driver/tty.py:
```diff
@@ -0,0 +1,26 @@
+    with subtest("failure"), no_sleep(), must_raise("action timed out"):
+        machine.wait_for_console_text("foobar", timeout=10)
+
+#     with subtest("success"):
```
Why is this commented out?

Synthetica9 replied:
Because wait_for_console_text did something else than I initially thought, and I haven't found a good way to test it yet.

marijanp commented on nixos/tests/nixos-test-driver/ocr.py:
```diff
@@ -0,0 +1,4 @@
+machine.wait_for_x()
+machine.succeed("xterm >&2 &")
+machine.wait_for_text("root@", timeout=10)
+# assert "root@" in machine.get_screen_text_variants()
```
Why is this commented out?

Synthetica9 replied:
Because it doesn't _quite_ work this way, I'll edit it to a variant that does work

roberth reviewed (commented):
This is awesome.

I've refactored the nixos test expressions to be module based, requiring a small rebase of the (small) `testing-python.nix` part. I hope my suggestion is useful.

roberth commented on nixos/tests/all-tests.nix:
```diff
@@ -355,6 +355,7 @@ in
   nixops = handleTest ./nixops/default.nix {};
   nixos-generate-config = handleTest ./nixos-generate-config.nix {};
   nixpkgs = pkgs.callPackage ../modules/misc/nixpkgs/test.nix { inherit evalMinimalConfig; };
+  nixos-test-driver = handleTest ./nixos-test-driver {};
```
I did the thing:

https://github.com/NixOS/nixpkgs/blob/563c3fb69742d2235f735aad75e3346f833ea720/nixos/tests/all-tests.nix#L87-L91

That means that the name is now "taken", but something like `nixos-test-driver.driver` should be fine.

RaitoBezarius replied:
Will wire it up in driver.

RaitoBezarius replied:
Wired up.

RaitoBezarius commented on nixos/lib/testing-python.nix:
```diff
@@ -52,16 +52,17 @@ rec {
     let
       # Reifies and correctly wraps the python test driver for
       # the respective qemu version and with or without ocr support
-      testDriver = pkgs.callPackage ./test-driver {
```
I partially tried something to rebase it, I still need to wire up the coverage driver and I think I will be good.

RaitoBezarius replied:
I think this is wired up, let me know what do you think.

RaitoBezarius commented on nixos/tests/nixos-test-driver/coverage-driver.nix:
```diff
@@ -0,0 +1,39 @@
+    COVERAGE_RCFILE=${coverage-rc} ${coverage} run -- "$@"
+    RETURN_CODE=$?
+
+    if whoami | grep "nixbld"; then
```
This is actually incorrect… In auto-uid-allocation, nixbld does not appear in the username.

RaitoBezarius commented on nixos/tests/nixos-test-driver/coverage-driver.nix:
```diff
@@ -0,0 +1,39 @@
+    sed "s@${driver}@$out@" ${driver}/bin/nixos-test-driver > $out/bin/nixos-test-driver
+    chmod +x $out/bin/nixos-test-driver
+    chmod +x $out/bin/.nixos-test-driver-wrapped
+    # exit 1
```
Not sure I understand the intent here.

RaitoBezarius committed:
nixos/lib/test-driver: passthru driver specific tests

Now, whenever we will modify the nixos-test-driver itself,
ofborg will run its self-test suite.

RaitoBezarius committed:
nixos/lib/test-driver: ruff-ify the driver.py

As it was complaining in some place, I had to `ruff`-ify + `black`-ify.

roberth commented on nixos/lib/testing/driver.nix:
```diff
@@ -108,6 +108,12 @@ in
 
+    makeTestDriver = mkOption {
+      description = mdDoc "Function that returns the package containing the test driver that will perform the testing instrumentation";
+      type = types.functionTo types.package;
```
```suggestion
      type = types.functionTo types.package;
      internal = true;
```

This does not seem like a principled interface that we want to support indefinitely.

(otherwise, needs way more docs)

RaitoBezarius replied:
Agreed

roberth commented on nixos/tests/nixos-test-driver/default.nix:
```diff
@@ -0,0 +1,75 @@
+let
+  lib = pkgs.lib;
+  machines = import ./machines.nix;
+  mkTest =
+    name:
+    { script
+    , nodes ? { inherit (machines) basic; }
+    , enableOCR ? false
+    , skipTypeCheck ? false
+    }:
+    import ../make-test-python.nix
```
This could be simplified with [`pkgs.testers.runNixOSTest`](https://nixos.org/manual/nixpkgs/unstable/#tester-runNixOSTest). Also evals faster because it reuses `pkgs`.

roberth commented on nixos/lib/test-driver/default.nix:
```diff
@@ -41,4 +42,8 @@ python3Packages.buildPythonApplication {
   '';
+
+  passthru.tests = {
+    inherit (nixosTests.nixos-test-driver) driver;
```
Should we add all of them?

RaitoBezarius replied:
All of them := all of the driver.*
All of them := nixosTests.nixos-test.driver.* ?

Either case, more tests is always good in my book and the test driver has suffered from several backlashes in the past due to excessively (good I guess) type checking and formatting.

If you suggest me to add all of them, will do so.

RaitoBezarius commented:
@ofborg test nixos-test-driver.driver

RaitoBezarius committed:
circom: init at 2.1.6

https://github.com/iden3/circom

RaitoBezarius commented:
@ofborg test nixos-test-driver.driver

roberth commented on nixos/lib/test-driver/default.nix:
```diff
@@ -46,4 +46,6 @@ python3Packages.buildPythonApplication {
     black --check --diff .
   '';
+
+  passthru.tests = nixosTests.nixos-test-driver.driver;
```
I was thinking:

```suggestion
  passthru.tests = nixosTests.nixos-test-driver;
```

(Resuming thread https://github.com/NixOS/nixpkgs/pull/157161#discussion_r1369010401)

RaitoBezarius replied:
*Ah*, why not, actually.
